File: src/setup_wizard.py
"""Setup Wizard — first-run installer UI with progress bar."""

import sys
import os
import json
import threading
import subprocess

import objc
from AppKit import (
    NSApplication,
    NSApp,
    NSWindow,
    NSView,
    NSTextField,
    NSSecureTextField,
    NSButton,
    NSProgressIndicator,
    NSColor,
    NSFont,
    NSScreen,
    NSMakeRect,
    NSObject,
    NSWindowStyleMaskTitled,
    NSWindowStyleMaskClosable,
    NSBackingStoreBuffered,
    NSTextAlignmentCenter,
    NSTextAlignmentLeft,
    NSLineBreakByWordWrapping,
    NSBezelStyleRounded,
    NSApplicationActivationPolicyRegular,
    NSProgressIndicatorStyleBar,
)

CONFIG_PATH = os.path.expanduser("~/.live-translator.json")
WIDTH = 480
HEIGHT = 360


class SetupWizard:
    """macOS native setup wizard with progress bar."""

    def __init__(self, resource_dir):
        self._dir = resource_dir
        self._window = None
        self._api_input = None
        self._progress = None
        self._status_label = None
        self._step = 0
        self._result = None  # "launch" or None

    def run(self):
        """Show wizard, return True if setup completed."""
        app = NSApplication.sharedApplication()
        app.setActivationPolicy_(NSApplicationActivationPolicyRegular)

        screen = NSScreen.mainScreen().frame()
        x = (screen.size.width - WIDTH) / 2
        y = (screen.size.height - HEIGHT) / 2

        self._window = NSWindow.alloc().initWithContentRect_styleMask_backing_defer_(
            NSMakeRect(x, y, WIDTH, HEIGHT),
            NSWindowStyleMaskTitled | NSWindowStyleMaskClosable,
            NSBackingStoreBuffered, False,
        )
        self._window.setTitle_("Live Translator — Setup")
        self._window.setBackgroundColor_(NSColor.colorWithRed_green_blue_alpha_(0.12, 0.12, 0.14, 1.0))

        self._show_step1()
        self._window.makeKeyAndOrderFront_(None)
        app.activateIgnoringOtherApps_(True)
        app.run()

        return self._result == "launch"

    def _clear(self):
        for sub in list(self._window.contentView().subviews()):
            sub.removeFromSuperview()

    def _label(self, frame, text, size=14, weight=0.4, color=None, align=NSTextAlignmentLeft, wrap=False):
        f = NSTextField.alloc().initWithFrame_(frame)
        f.setStringValue_(text)
        f.setBezeled_(False)
        f.setDrawsBackground_(False)
        f.setEditable_(False)
        f.setSelectable_(False)
        f.setTextColor_(color or NSColor.whiteColor())
        f.setFont_(NSFont.systemFontOfSize_weight_(size, weight))
        f.setAlignment_(align)
        if wrap:
            f.setLineBreakMode_(NSLineBreakByWordWrapping)
            f.setMaximumNumberOfLines_(0)
        return f

    # ── Step 1: Welcome + API Key ──
    def _show_step1(self):
        self._clear()
        cv = self._window.contentView()

        # Step indicator
        cv.addSubview_(self._label(
            NSMakeRect(0, HEIGHT - 40, WIDTH, 20),
            "Step 1 of 3", 11, 0.3,
            NSColor.colorWithWhite_alpha_(0.4, 1.0),
            NSTextAlignmentCenter,
        ))

        # Title
        cv.addSubview_(self._label(
            NSMakeRect(30, HEIGHT - 80, WIDTH - 60, 30),
            "Welcome to Live Translator", 22, 0.6,
            NSColor.whiteColor(), NSTextAlignmentCenter,
        ))

        # Description
        cv.addSubview_(self._label(
            NSMakeRect(40, HEIGHT - 150, WIDTH - 80, 50),
            "Translate any audio playing on your Mac in real-time.\nEnter your OpenAI API key to get started.",
            13, 0.3, NSColor.colorWithWhite_alpha_(0.6, 1.0),
            NSTextAlignmentCenter, True,
        ))

        # API Key label
        cv.addSubview_(self._label(
            NSMakeRect(40, HEIGHT - 185, 200, 20),
            "OpenAI API Key", 12, 0.5,
            NSColor.colorWithWhite_alpha_(0.7, 1.0),
        ))

        # API Key input
        self._api_input = NSSecureTextField.alloc().initWithFrame_(
            NSMakeRect(40, HEIGHT - 215, WIDTH - 80, 28)
        )
        self._api_input.setPlaceholderString_("sk-...")
        self._api_input.setFont_(NSFont.monospacedSystemFontOfSize_weight_(13, 0.3))
        self._api_input.setTextColor_(NSColor.whiteColor())
        self._api_input.setBackgroundColor_(NSColor.colorWithRed_green_blue_alpha_(0.18, 0.18, 0.20, 1.0))
        self._api_input.setBezeled_(True)
        cv.addSubview_(self._api_input)

        # Link
        cv.addSubview_(self._label(
            NSMakeRect(40, HEIGHT - 240, WIDTH - 80, 16),
            "Get your key at platform.openai.com", 11, 0.2,
            NSColor.colorWithWhite_alpha_(0.35, 1.0),
        ))

        # Next button
        btn = NSButton.alloc().initWithFrame_(NSMakeRect(WIDTH - 140, 20, 100, 36))
        btn.setTitle_("Next →")
        btn.setBezelStyle_(NSBezelStyleRounded)
        btn.setTarget_(self)
        btn.setAction_(objc.selector(self._on_next1, signature=b"v@:@"))
        cv.addSubview_(btn)

        # Quit button
        qbtn = NSButton.alloc().initWithFrame_(NSMakeRect(40, 20, 80, 36))
        qbtn.setTitle_("Quit")
        qbtn.setBezelStyle_(NSBezelStyleRounded)
        qbtn.setTarget_(self)
        qbtn.setAction_(objc.selector(self._on_quit, signature=b"v@:@"))
        cv.addSubview_(qbtn)

    def _on_next1(self, sender):
        key = self._api_input.stringValue()
        if not key or len(key) < 10:
            # Error
            self._api_input.setTextColor_(NSColor.redColor())
            return

        # Save config
        config = {
            "openai_api_key": key,
            "source_locale": "en-US",
            "target_lang": "tr",
            "model": "gpt-4o-mini",
            "tts_provider": "piper",
            "tts_voice": "nova",
            "tts_speed": 1.0,
        }
        with open(CONFIG_PATH, "w") as f:
            json.dump(config, f, indent=2)

        self._show_step2()

    # ── Step 2: Installing ──
    def _show_step2(self):
        self._clear()
        cv = self._window.contentView()

        cv.addSubview_(self._label(
            NSMakeRect(0, HEIGHT - 40, WIDTH, 20),
            "Step 2 of 3", 11, 0.3,
            NSColor.colorWithWhite_alpha_(0.4, 1.0),
            NSTextAlignmentCenter,
        ))

        cv.addSubview_(self._label(
            NSMakeRect(30, HEIGHT - 80, WIDTH - 60, 30),
            "Installing...", 22, 0.6,
            NSColor.whiteColor(), NSTextAlignmentCenter,
        ))

        cv.addSubview_(self._label(
            NSMakeRect(40, HEIGHT - 120, WIDTH - 80, 40),
            "Setting up Python environment and downloading packages.\nThis only happens once.",
            13, 0.3, NSColor.colorWithWhite_alpha_(0.5, 1.0),
            NSTextAlignmentCenter, True,
        ))

        # Progress bar
        self._progress = NSProgressIndicator.alloc().initWithFrame_(
            NSMakeRect(40, HEIGHT - 165, WIDTH - 80, 20)
        )
        self._progress.setStyle_(NSProgressIndicatorStyleBar)
        self._progress.setMinValue_(0)
        self._progress.setMaxValue_(100)
        self._progress.setDoubleValue_(0)
        self._progress.setIndeterminate_(False)
        cv.addSubview_(self._progress)

        # Status label
        self._status_label = self._label(
            NSMakeRect(40, HEIGHT - 195, WIDTH - 80, 20),
            "Preparing...", 12, 0.3,
            NSColor.colorWithWhite_alpha_(0.45, 1.0),
            NSTextAlignmentCenter,
        )
        cv.addSubview_(self._status_label)

        # Start install in background
        threading.Thread(target=self._do_install, daemon=True).start()

    def _update_progress(self, value, text):
        from PyObjCTools import AppHelper
        def update():
            if self._progress:
                self._progress.setDoubleValue_(value)
            if self._status_label:
                self._status_label.setStringValue_(text)
        AppHelper.callAfter(update)

    def _do_install(self):
        venv = os.path.join(self._dir, ".venv")
        log = os.path.expanduser("~/Library/Logs/LiveTranslator.log")

        try:
            # Step: venv
            self._update_progress(10, "Creating Python environment...")
            subprocess.run(
                ["python3", "-m", "venv", venv],
                capture_output=True, timeout=60,
            )

            # Step: pip install
            self._update_progress(25, "Installing packages (this may take a minute)...")
            pip = os.path.join(venv, "bin", "pip")
            req = os.path.join(self._dir, "requirements.txt")
            subprocess.run(
                [pip, "install", "-q", "-r", req],
                capture_output=True, timeout=300,
            )

            # Step: TTS models
            self._update_progress(70, "Downloading voice models...")
            dl_script = os.path.join(self._dir, "scripts", "download_models.sh")
            if os.path.exists(dl_script):
                subprocess.run(
                    ["bash", dl_script],
                    capture_output=True, timeout=600,
                    cwd=self._dir,
                )

            self._update_progress(100, "Done!")

            # Show step 3
            from PyObjCTools import AppHelper
            AppHelper.callAfter(self._show_step3)

        except Exception as e:
            self._update_progress(0, f"Error: {e}")

    # ── Step 3: Ready ──
    def _show_step3(self):
        self._clear()
        cv = self._window.contentView()

        cv.addSubview_(self._label(
            NSMakeRect(0, HEIGHT - 40, WIDTH, 20),
            "Step 3 of 3", 11, 0.3,
            NSColor.colorWithWhite_alpha_(0.4, 1.0),
            NSTextAlignmentCenter,
        ))

        cv.addSubview_(self._label(
            NSMakeRect(30, HEIGHT - 85, WIDTH - 60, 30),
            "Ready!", 26, 0.6,
            NSColor.colorWithRed_green_blue_alpha_(0.4, 0.8, 0.5, 1.0),
            NSTextAlignmentCenter,
        ))

        cv.addSubview_(self._label(
            NSMakeRect(40, HEIGHT - 160, WIDTH - 80, 60),
            "Live Translator is ready to use.\n\nPlay any audio on your Mac and translations\nwill appear in the floating panel.",
            13, 0.3, NSColor.colorWithWhite_alpha_(0.6, 1.0),
            NSTextAlignmentCenter, True,
        ))

        # Launch button
        btn = NSButton.alloc().initWithFrame_(NSMakeRect((WIDTH - 140) / 2, 30, 140, 40))
        btn.setTitle_("Launch")
        btn.setBezelStyle_(NSBezelStyleRounded)
        btn.setFont_(NSFont.systemFontOfSize_weight_(15, 0.5))
        btn.setTarget_(self)
        btn.setAction_(objc.selector(self._on_launch, signature=b"v@:@"))
        cv.addSubview_(btn)

    def _on_launch(self, sender):
        self._result = "launch"
        NSApp.stop_(None)
        # Post dummy event to unblock run loop
        from AppKit import NSEvent, NSEventTypeApplicationDefined
        e = NSEvent.otherEventWithType_location_modifierFlags_timestamp_windowNumber_context_subtype_data1_data2_(
            NSEventTypeApplicationDefined, (0, 0), 0, 0, 0, None, 0, 0, 0
        )
        NSApp.postEvent_atStart_(e, True)

    def _on_quit(self, sender):
        self._result = None
        NSApp.stop_(None)
        from AppKit import NSEvent, NSEventTypeApplicationDefined
        e = NSEvent.otherEventWithType_location_modifierFlags_timestamp_windowNumber_context_subtype_data1_data2_(
            NSEventTypeApplicationDefined, (0, 0), 0, 0, 0, None, 0, 0, 0
        )
        NSApp.postEvent_atStart_(e, True)


def needs_setup(resource_dir):
    """Check if first-run setup is needed."""
    config_exists = os.path.exists(CONFIG_PATH)
    venv_exists = os.path.exists(os.path.join(resource_dir, ".venv"))

    if not config_exists:
        return True

    try:
        with open(CONFIG_PATH) as f:
            config = json.load(f)
        if not config.get("openai_api_key"):
            return True
    except Exception:
        return True

    if not venv_exists:
        return True

    return False


if __name__ == "__main__":
    res_dir = sys.argv[1] if len(sys.argv) > 1 else os.path.dirname(__file__)
    wizard = SetupWizard(res_dir)
    if wizard.run():
        print("LAUNCH")
    else:
        print("QUIT")

File: scripts/setup_wizard.swift
import Cocoa
import QuartzCore

let W: CGFloat = 540
let H: CGFloat = 460

class SetupWindow: NSWindow, NSTextFieldDelegate {
    var apiKeyField: NSTextField!
    var progressBar: NSProgressIndicator!
    var statusLabel: NSTextField!
    var resourceDir: String = ""
    var appIcon: NSImage?
    var hasKey: Bool = false

    func setup(resDir: String) {
        resourceDir = resDir
        setContentSize(NSSize(width: W, height: H))
        styleMask = [.titled, .closable, .fullSizeContentView]
        titlebarAppearsTransparent = true
        titleVisibility = .hidden
        isMovableByWindowBackground = true
        backgroundColor = NSColor(red: 0.10, green: 0.10, blue: 0.12, alpha: 1)
        center()

        // App icon
        let iconPath = resDir + "/AppIcon.icns"
        if FileManager.default.fileExists(atPath: iconPath) {
            appIcon = NSImage(contentsOfFile: iconPath)
        }

        // Enable paste (Cmd+V) — add Edit menu
        let mainMenu = NSMenu()
        let editMenu = NSMenu(title: "Edit")
        editMenu.addItem(withTitle: "Cut", action: #selector(NSText.cut(_:)), keyEquivalent: "x")
        editMenu.addItem(withTitle: "Copy", action: #selector(NSText.copy(_:)), keyEquivalent: "c")
        editMenu.addItem(withTitle: "Paste", action: #selector(NSText.paste(_:)), keyEquivalent: "v")
        editMenu.addItem(withTitle: "Select All", action: #selector(NSText.selectAll(_:)), keyEquivalent: "a")
        let editMenuItem = NSMenuItem(title: "Edit", action: nil, keyEquivalent: "")
        editMenuItem.submenu = editMenu
        mainMenu.addItem(editMenuItem)
        NSApp.mainMenu = mainMenu

        // Check if key already exists
        let configPath = NSHomeDirectory() + "/.live-translator.json"
        if let data = try? Data(contentsOf: URL(fileURLWithPath: configPath)),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let key = json["openai_api_key"] as? String, !key.isEmpty {
            hasKey = true
            showStep2()  // Key exists, skip to installation
        } else {
            showStep1()
        }
    }

    func clear() { contentView?.subviews.forEach { $0.removeFromSuperview() } }

    // ─── Helpers ───

    func lbl(_ frame: NSRect, _ text: String, size: CGFloat = 14, weight: NSFont.Weight = .regular,
             color: NSColor = .white, align: NSTextAlignment = .center) -> NSTextField {
        let l = NSTextField(frame: frame)
        l.stringValue = text; l.isBezeled = false; l.drawsBackground = false
        l.isEditable = false; l.isSelectable = false; l.textColor = color
        l.font = NSFont.systemFont(ofSize: size, weight: weight)
        l.alignment = align; l.lineBreakMode = .byWordWrapping; l.maximumNumberOfLines = 0
        return l
    }

    func stepBadge(_ step: Int, total: Int = 3) -> NSView {
        let v = NSView(frame: NSRect(x: (W-90)/2, y: H-48, width: 90, height: 24))
        v.wantsLayer = true
        v.layer?.backgroundColor = NSColor(white: 1, alpha: 0.05).cgColor
        v.layer?.cornerRadius = 12
        v.layer?.borderWidth = 0.5
        v.layer?.borderColor = NSColor(white: 1, alpha: 0.1).cgColor
        v.addSubview(lbl(NSRect(x: 0, y: 2, width: 90, height: 18),
                         "STEP \(step) / \(total)", size: 10, weight: .bold,
                         color: NSColor(white: 0.45, alpha: 1)))
        return v
    }

    func appIconView(_ y: CGFloat, size: CGFloat = 64) -> NSImageView {
        let iv = NSImageView(frame: NSRect(x: (W-size)/2, y: y, width: size, height: size))
        if let icon = appIcon { iv.image = icon }
        iv.imageScaling = .scaleProportionallyUpOrDown
        iv.wantsLayer = true
        iv.layer?.cornerRadius = size * 0.22
        iv.layer?.masksToBounds = true
        iv.layer?.borderWidth = 1
        iv.layer?.borderColor = NSColor(white: 1, alpha: 0.08).cgColor
        return iv
    }

    func sep(_ y: CGFloat) -> NSView {
        let v = NSView(frame: NSRect(x: 50, y: y, width: W-100, height: 1))
        v.wantsLayer = true
        v.layer?.backgroundColor = NSColor(white: 1, alpha: 0.06).cgColor
        return v
    }

    func greenBtn(_ frame: NSRect, _ title: String, action: Selector) -> NSView {
        let bg = NSView(frame: frame)
        bg.wantsLayer = true
        bg.layer?.backgroundColor = NSColor(red: 0.22, green: 0.58, blue: 0.36, alpha: 1).cgColor
        bg.layer?.cornerRadius = frame.height / 2
        let btn = NSButton(frame: NSRect(x: 0, y: 0, width: frame.width, height: frame.height))
        btn.title = title; btn.isBordered = false
        btn.font = NSFont.systemFont(ofSize: 14, weight: .semibold)
        btn.contentTintColor = .white; btn.target = self; btn.action = action
        bg.addSubview(btn)
        return bg
    }

    func ghostBtn(_ frame: NSRect, _ title: String, action: Selector) -> NSButton {
        let btn = NSButton(frame: frame)
        btn.title = title; btn.isBordered = false
        btn.font = NSFont.systemFont(ofSize: 13, weight: .regular)
        btn.contentTintColor = NSColor(white: 0.4, alpha: 1)
        btn.target = self; btn.action = action
        return btn
    }

    // ═══════════════════════════════════════
    // MARK: Step 1 — API Key
    // ═══════════════════════════════════════
    func showStep1() {
        clear()
        guard let cv = contentView else { return }

        cv.addSubview(stepBadge(1))
        cv.addSubview(appIconView(H - 130))

        cv.addSubview(lbl(NSRect(x: 30, y: H-180, width: W-60, height: 35),
                          "Welcome to Live Translator", size: 24, weight: .bold))

        cv.addSubview(lbl(NSRect(x: 50, y: H-225, width: W-100, height: 34),
                          "Translate any audio playing on your Mac in real-time.\nEnter your OpenAI API key to get started.",
                          size: 13, weight: .regular, color: NSColor(white: 0.45, alpha: 1)))

        cv.addSubview(sep(H - 245))

        // Label
        cv.addSubview(lbl(NSRect(x: 55, y: H-275, width: 200, height: 18),
                          "OpenAI API Key", size: 12, weight: .semibold,
                          color: NSColor(white: 0.55, alpha: 1), align: .left))

        // Input — regular NSTextField (not secure) so paste works naturally
        apiKeyField = NSTextField(frame: NSRect(x: 52, y: H-315, width: W-104, height: 32))
        apiKeyField.placeholderString = "sk-proj-..."
        apiKeyField.font = NSFont.monospacedSystemFont(ofSize: 13, weight: .regular)
        apiKeyField.textColor = NSColor(white: 0.85, alpha: 1)
        apiKeyField.backgroundColor = NSColor(red: 0.14, green: 0.14, blue: 0.16, alpha: 1)
        apiKeyField.isBezeled = true
        apiKeyField.bezelStyle = .roundedBezel
        apiKeyField.focusRingType = .none
        apiKeyField.wantsLayer = true
        apiKeyField.layer?.cornerRadius = 8
        apiKeyField.layer?.borderWidth = 1
        apiKeyField.layer?.borderColor = NSColor(white: 1, alpha: 0.1).cgColor
        cv.addSubview(apiKeyField)

        // Link
        cv.addSubview(lbl(NSRect(x: 55, y: H-340, width: W-110, height: 15),
                          "Get your key at platform.openai.com →", size: 11, weight: .regular,
                          color: NSColor(red: 0.35, green: 0.65, blue: 0.5, alpha: 0.8), align: .left))

        // Buttons
        cv.addSubview(ghostBtn(NSRect(x: 45, y: 25, width: 70, height: 36), "Quit", action: #selector(onQuit)))
        cv.addSubview(greenBtn(NSRect(x: W-170, y: 22, width: 130, height: 42), "Next →", action: #selector(onNext1)))
    }

    @objc func onNext1() {
        let key = apiKeyField.stringValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard key.count >= 10 else {
            // Shake
            let cx = apiKeyField.frame.origin.x + apiKeyField.frame.width / 2
            let anim = CAKeyframeAnimation(keyPath: "position.x")
            anim.values = [cx, cx-10, cx+10, cx-5, cx+5, cx] as [NSNumber]
            anim.duration = 0.4
            apiKeyField.layer?.add(anim, forKey: "shake")
            apiKeyField.layer?.borderColor = NSColor.red.withAlphaComponent(0.5).cgColor
            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                self.apiKeyField.layer?.borderColor = NSColor(white: 1, alpha: 0.1).cgColor
            }
            return
        }

        // Save config
        let config: [String: Any] = [
            "openai_api_key": key, "source_locale": "en-US", "target_lang": "tr",
            "model": "gpt-4o-mini", "tts_provider": "piper", "tts_voice": "nova", "tts_speed": 1.0
        ]
        if let data = try? JSONSerialization.data(withJSONObject: config, options: .prettyPrinted) {
            try? data.write(to: URL(fileURLWithPath: NSHomeDirectory() + "/.live-translator.json"))
        }
        showStep2()
    }

    // ═══════════════════════════════════════
    // MARK: Step 2 — Installing
    // ═══════════════════════════════════════
    func showStep2() {
        clear()
        guard let cv = contentView else { return }

        cv.addSubview(stepBadge(2))
        cv.addSubview(appIconView(H - 130))

        cv.addSubview(lbl(NSRect(x: 30, y: H-180, width: W-60, height: 35),
                          "Setting up...", size: 24, weight: .bold))

        cv.addSubview(lbl(NSRect(x: 50, y: H-215, width: W-100, height: 30),
                          "Installing packages and downloading voice models.\nThis only happens once.",
                          size: 13, weight: .regular, color: NSColor(white: 0.42, alpha: 1)))

        cv.addSubview(sep(H - 235))

        // Progress track
        let trackBg = NSView(frame: NSRect(x: 55, y: H-268, width: W-110, height: 8))
        trackBg.wantsLayer = true
        trackBg.layer?.backgroundColor = NSColor(white: 1, alpha: 0.05).cgColor
        trackBg.layer?.cornerRadius = 4
        cv.addSubview(trackBg)

        progressBar = NSProgressIndicator(frame: NSRect(x: 55, y: H-268, width: W-110, height: 8))
        progressBar.style = .bar; progressBar.minValue = 0; progressBar.maxValue = 100
        progressBar.doubleValue = 0; progressBar.isIndeterminate = false
        progressBar.wantsLayer = true; progressBar.layer?.cornerRadius = 4
        cv.addSubview(progressBar)

        statusLabel = lbl(NSRect(x: 55, y: H-292, width: W-110, height: 16),
                          "Preparing...", size: 11, weight: .medium,
                          color: NSColor(white: 0.38, alpha: 1))
        cv.addSubview(statusLabel)

        // Checklist
        let items = ["Python environment", "Required packages", "Voice models (~580MB)"]
        for (i, item) in items.enumerated() {
            let y = H - CGFloat(328 + i * 32)
            let dot = lbl(NSRect(x: 70, y: y, width: 20, height: 22), "○",
                          size: 16, weight: .ultraLight, color: NSColor(white: 0.22, alpha: 1))
            dot.tag = 100 + i
            cv.addSubview(dot)
            let text = lbl(NSRect(x: 95, y: y + 1, width: W-160, height: 20), item,
                           size: 13, weight: .regular, color: NSColor(white: 0.38, alpha: 1), align: .left)
            text.tag = 200 + i
            cv.addSubview(text)
        }

        DispatchQueue.global().async { self.doInstall() }
    }

    func setProgress(_ val: Double, _ text: String) {
        DispatchQueue.main.async {
            self.progressBar?.doubleValue = val
            self.statusLabel?.stringValue = text
        }
    }

    func checkItem(_ index: Int) {
        DispatchQueue.main.async {
            guard let cv = self.contentView else { return }
            if let dot = cv.viewWithTag(100 + index) as? NSTextField {
                dot.stringValue = "✓"
                dot.textColor = NSColor(red: 0.35, green: 0.78, blue: 0.48, alpha: 1)
                dot.font = NSFont.systemFont(ofSize: 14, weight: .bold)
            }
            if let text = cv.viewWithTag(200 + index) as? NSTextField {
                text.textColor = NSColor(white: 0.6, alpha: 1)
            }
        }
    }

    func doInstall() {
        let supportDir = NSHomeDirectory() + "/Library/Application Support/LiveTranslator"
        try? FileManager.default.createDirectory(atPath: supportDir, withIntermediateDirectories: true)
        let venv = supportDir + "/venv"
        let log = NSHomeDirectory() + "/Library/Logs/LiveTranslator.log"

        // Find python3
        let pythonPaths = ["/opt/homebrew/bin/python3", "/usr/local/bin/python3", "/usr/bin/python3"]
        var python = pythonPaths.first { FileManager.default.fileExists(atPath: $0) }

        // No python found — install via Homebrew
        if python == nil {
            setProgress(2, "Python not found, installing...")

            // Check if Homebrew exists
            let brewPaths = ["/opt/homebrew/bin/brew", "/usr/local/bin/brew"]
            let brew = brewPaths.first { FileManager.default.fileExists(atPath: $0) }

            if let brew = brew {
                // Install Python via Homebrew
                shell([brew, "install", "python@3.13"], log)
                python = "/opt/homebrew/bin/python3"
                if !FileManager.default.fileExists(atPath: python!) {
                    python = "/usr/local/bin/python3"
                }
            } else {
                // Install Homebrew first, then Python
                setProgress(1, "Installing Homebrew...")
                shell(["/bin/bash", "-c",
                       "NONINTERACTIVE=1 /bin/bash -c \"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)\""],
                      log)

                let newBrew = brewPaths.first { FileManager.default.fileExists(atPath: $0) }
                if let newBrew = newBrew {
                    setProgress(2, "Installing Python...")
                    shell([newBrew, "install", "python@3.13"], log)
                    python = "/opt/homebrew/bin/python3"
                }
            }

            // Final check
            if python == nil || !FileManager.default.fileExists(atPath: python!) {
                DispatchQueue.main.async {
                    self.showError("Python 3.11+ is required but could not be installed.\n\nPlease install manually:\n  brew install python\n\nOr download from python.org")
                }
                return
            }
        }

        let pythonBin = python!

        setProgress(5, "Creating Python environment...")
        shell([pythonBin, "-m", "venv", venv], log)
        checkItem(0)
        setProgress(20, "Environment ready")
        Thread.sleep(forTimeInterval: 0.3)

        setProgress(25, "Installing packages...")
        let pip = venv + "/bin/pip"
        let req = resourceDir + "/requirements.txt"
        shell([pip, "install", "-q", "-r", req], log)
        checkItem(1)
        setProgress(60, "Packages installed")
        Thread.sleep(forTimeInterval: 0.3)

        setProgress(65, "Downloading voice models...")
        let dl = resourceDir + "/scripts/download_models.sh"
        if FileManager.default.fileExists(atPath: dl) {
            shell(["/bin/bash", dl], log, cwd: resourceDir)
        }
        checkItem(2)
        setProgress(100, "All done!")

        Thread.sleep(forTimeInterval: 0.5)
        DispatchQueue.main.async { self.showStep3() }
    }

    func shell(_ args: [String], _ log: String, cwd: String? = nil) {
        // Use /bin/bash -c with properly quoted args — most reliable
        let quoted = args.map { "'\($0.replacingOccurrences(of: "'", with: "'\\''"))'" }.joined(separator: " ")
        var cmd = quoted
        if let dir = cwd {
            cmd = "cd '\(dir)' && \(quoted)"
        }
        cmd += " >> '\(log)' 2>&1"

        let t = Process()
        t.launchPath = "/bin/bash"
        t.arguments = ["-c", cmd]
        do {
            try t.run()
            t.waitUntilExit()
        } catch {
            // Log error
            let errMsg = "Shell error: \(error)\nCommand: \(cmd)\n"
            if let data = errMsg.data(using: .utf8) {
                let fh = FileHandle(forWritingAtPath: log) ?? {
                    FileManager.default.createFile(atPath: log, contents: nil)
                    return FileHandle(forWritingAtPath: log)!
                }()
                fh.seekToEndOfFile()
                fh.write(data)
                fh.closeFile()
            }
        }
    }

    // ═══════════════════════════════════════
    // MARK: Step 3 — Ready
    // ═══════════════════════════════════════
    func showStep3() {
        clear()
        guard let cv = contentView else { return }

        cv.addSubview(stepBadge(3))
        cv.addSubview(appIconView(H - 135, size: 72))

        cv.addSubview(lbl(NSRect(x: 30, y: H-190, width: W-60, height: 38),
                          "You're all set!", size: 28, weight: .bold,
                          color: NSColor(red: 0.4, green: 0.82, blue: 0.52, alpha: 1)))

        cv.addSubview(sep(H - 205))

        cv.addSubview(lbl(NSRect(x: 50, y: H-280, width: W-100, height: 60),
                          "Live Translator is ready.\n\nPlay any audio on your Mac and translations\nwill appear in the floating panel.\n\nYou can change settings anytime from the panel.",
                          size: 13, weight: .regular, color: NSColor(white: 0.48, alpha: 1)))

        cv.addSubview(greenBtn(NSRect(x: (W-170)/2, y: 30, width: 170, height: 46),
                               "Launch →", action: #selector(onLaunch)))
    }

    func showError(_ message: String) {
        clear()
        guard let cv = contentView else { return }

        cv.addSubview(appIconView(H - 130))
        cv.addSubview(lbl(NSRect(x: 30, y: H-175, width: W-60, height: 35),
                          "Setup Failed", size: 24, weight: .bold,
                          color: NSColor(red: 0.9, green: 0.3, blue: 0.3, alpha: 1)))
        cv.addSubview(sep(H - 190))
        cv.addSubview(lbl(NSRect(x: 50, y: H-290, width: W-100, height: 80),
                          message, size: 13, weight: .regular,
                          color: NSColor(white: 0.5, alpha: 1)))
        cv.addSubview(ghostBtn(NSRect(x: (W-80)/2, y: 30, width: 80, height: 36),
                               "Quit", action: #selector(onQuit)))
    }

    @objc func onLaunch() {
        // Write marker file — launcher reads this
        let marker = "/tmp/.livetranslator_launch"
        try? "LAUNCH".write(toFile: marker, atomically: true, encoding: .utf8)
        // Verify
        if FileManager.default.fileExists(atPath: marker) {
            exit(0)
        }
        exit(0)
    }

    @objc func onQuit() {
        exit(1)
    }
}

// ═══ Main ═══
let args = CommandLine.arguments
let resDir = args.count > 1 ? args[1] : FileManager.default.currentDirectoryPath
let configPath = NSHomeDirectory() + "/.live-translator.json"
let venvPath = NSHomeDirectory() + "/Library/Application Support/LiveTranslator/venv"

// Check if BOTH config and venv exist
var needsSetup = true
if FileManager.default.fileExists(atPath: configPath),
   FileManager.default.fileExists(atPath: venvPath + "/bin/python"),
   let data = try? Data(contentsOf: URL(fileURLWithPath: configPath)),
   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
   let key = json["openai_api_key"] as? String, !key.isEmpty {
    needsSetup = false
}

if !needsSetup { print("LAUNCH"); exit(0) }

let app = NSApplication.shared
app.setActivationPolicy(.regular)
let win = SetupWindow(contentRect: .zero, styleMask: [.titled, .closable, .fullSizeContentView],
                      backing: .buffered, defer: false)
win.setup(resDir: resDir)
win.makeKeyAndOrderFront(nil)
app.activate(ignoringOtherApps: true)
app.run()
